prometheus-server-metrics/src/main/scala/org/http4s/server/prometheus/PrometheusMetrics.scala

```diff
@@ -0,0 +1,240 @@
+            serviceMetrics.generalMetrics.activeRequests
+          )(m, start, finalBodyTime)
+          _ <- responseMetrics(
+            serviceMetrics.responseHist,
+            r.status,
+            start,
+            finalBodyTime
+          )
+        } yield ()
+      }
+      .handleErrorWith(e =>
+        for {
+          terminationTime <- Stream.eval(Sync[F].delay(System.nanoTime))
+          _ <- Stream.eval(
+            Sync[F].delay(
+              serviceMetrics.generalMetrics.abnormalTerminations
+                .observe(SimpleTimer.elapsedSecondsFromNanos(start, terminationTime))
+            ))
+          result <- Stream.raiseError[Byte](e).covary[F]
+        } yield result)
+    r.copy(body = newBody)
+  }
+
+  private def onServiceError[F[_]: Sync](m: Method, begin: Long, sm: ServiceMetrics): F[Unit] =
+    for {
+      end <- Sync[F].delay(System.nanoTime)
+      _ <- requestMetrics(
+        sm.requestHist,
+        sm.generalMetrics.totalRequests,
+        sm.generalMetrics.activeRequests
+      )(m, begin, end)
+      _ <- Sync[F].delay(
+        sm.generalMetrics.serviceErrors.observe(SimpleTimer.elapsedSecondsFromNanos(begin, end)))
+    } yield ()
+
+  private def requestMetrics[F[_]: Sync](
+      rh: RequestHistograms,
+      totalReqs: Histogram,
+      activeReqs: Gauge)(m: Method, begin: Long, end: Long): F[Unit] = Sync[F].delay {
+    val histogram = requestHistogram(rh, m)
+    val elapsed = SimpleTimer.elapsedSecondsFromNanos(begin, end)
+    histogram.observe(elapsed)
+    totalReqs.observe(elapsed)
+    activeReqs.dec()
+  }
+
+  private def responseMetrics[F[_]: Sync](
+      rh: ResponseHistograms,
+      status: Status,
+      begin: Long,
+      end: Long): F[Unit] = Sync[F].delay {
+    responseHistogram(rh, status).observe(SimpleTimer.elapsedSecondsFromNanos(begin, end))
+  }
+
+  private def requestHistogram[F[_]: Sync](rt: RequestHistograms, method: Method): Histogram =
+    method match {
+      case Method.GET => rt.getReq
+      case Method.POST => rt.postReq
+      case Method.PUT => rt.putReq
+      case Method.HEAD => rt.headReq
+      case Method.MOVE => rt.moveReq
+      case Method.OPTIONS => rt.optionsReq
+      case Method.TRACE => rt.traceReq
+      case Method.CONNECT => rt.connectReq
+      case Method.DELETE => rt.deleteReq
+      case _ => rt.otherReq
+    }
+
+  private def responseHistogram(responseTimers: ResponseHistograms, status: Status): Histogram =
+    status.code match {
+      case hundreds if hundreds < 200 => responseTimers.resp1xx
+      case twohundreds if twohundreds < 300 => responseTimers.resp2xx
+      case threehundreds if threehundreds < 400 => responseTimers.resp3xx
+      case fourhundreds if fourhundreds < 500 => responseTimers.resp4xx
+      case _ => responseTimers.resp5xx
+    }
+
+  def apply[F[_]: Effect](
+      prefix: String = "org_http4s_server",
+      c: CollectorRegistry): F[HttpMiddleware[F]] = Sync[F].delay {
```

Comment:
How about `HttpService[F] => F[HttpService[F]]`?  Does the whole function need to be in an effect, or is just effectful function?  And then does `Kleisli[F, HttpService[F], HttpService[F]]` improve the ergonomics?  I think this is our first effectful middleware?